docs: document createSpy with examples (#1138)
posva/pinia: packages/docs/cookbook/testing.md

## Unit testing components

> This can be achieved with `createTestingPinia()`, which returns a pinia instance designed for unit tests.

Reviewer: ```suggestion
This can be achieved with `createTestingPinia()`, which returns a pinia instance designed to help unit tests components.
```

store.$patch({ name: 'new name' })
expect(store.name).toBe('new name')

> // actions are stubbed by default, so this call won't actually
> // execute the implementation defined by the store. See below 
> // to customize this behavior.

Reviewer: ```suggestion
// actions are stubbed by default, meaning they don't execute their code by default.
// See below to customize this behavior.
```

store.someAction()

expect(store.someAction).toHaveBeenCalledTimes(1)
expect(store.someAction).toHaveBeenLastCalledWith()
```

Please note that if you are using Vue 2, `@vue/test-utils` requires a [slightly different configuration](#unit-test-components-vue-2).

### Customizing behavior of actions

> By default, `createTestingPinia` stubs out all store actions, such that they are not executed when called during unit tests. This allows you to test your components and stores separately.

Reviewer: ```suggestion
`createTestingPinia` stubs out all store actions unless told otherwise. This allows you to test your components and stores separately.
```

> If you want your actions to execute during unit tests, specify `stubActions: true` as an option when calling `createTestingPinia`:

Reviewer: ```suggestion
If you want to revert this behavior and normally execute your actions during tests, specify `stubActions: false` when calling `createTestingPinia`:
```

> ```

Reviewer: ```suggestion
```js
```

const wrapper = mount(Counter, {
  global: {
    plugins: [createTestingPinia({ stubActions: false })],
  },
})

const store = useSomeStore()

// Now this call WILL execute the implementation defined by the store
store.someAction()

// ...but it's still wrapped with a spy, so you can inspect calls 
expect(store.someAction).toHaveBeenCalledTimes(1)

```

### Specifying the createSpy function

> When using Jest, or vitest with `globals: true`, `createTestingPinia` will create action stubs automatically using the spy function defined by the test framework (`jest.fn` or `vitest.fn`). If you are using a different framework, you'll need to provide a [createSpy](https://pinia.vuejs.org/api/interfaces/pinia_testing.testingoptions.html#createspy) and supply it as a parameter.

Reviewer: ```suggestion
When using Jest, or vitest with `globals: true`, `createTestingPinia` automatically stubs actions using the spy function based on the existing test framework (`jest.fn` or `vitest.fn`). If you are using a different framework, you'll need to provide a [createSpy](/api/interfaces/pinia_testing.testingoptions.html#createspy) option:
```

> ```
> import sinon from 'sinon'
> createTestingPinia({ 
>   createSpy: sinon.spy // use sinon's spy to wrap actions
>   })
> ```

Reviewer: ```suggestion
```js
import sinon from 'sinon'

createTestingPinia({ 
  createSpy: sinon.spy // use sinon's spy to wrap actions
})
```
```

> 
> (Note that the provided `createSpy` will also be used to wrap the `$patch` method of the store).
> 

Reviewer: ```suggestion
```

expect(store.someAction).toHaveBeenLastCalledWith()
```

> If you are not using Jest (for example, you are using vitest), you will need to provide a [createSpy](https://pinia.vuejs.org/api/interfaces/pinia_testing.testingoptions.html#createspy) to `createTestingPinia()`, like:

Reviewer: It actually works if you have `globals: true` with vitest. Let's:

- Add a new subsection about actions being mocked
- Say they are automatically mocked in jest and in vite with globals: true
- Show how to add your own createSpy function

Author: Hmm, I have this is my vite.config.ts:

```
    test: {
      globals: true,
      environment: 'happy-dom',
    } as VitestUserConfig,
```

let me investigate.

Author: @posva I think a new version of @pinai/testing needs to created and published. The behavior with `vi` was changed/fixed in [this commit](https://github.com/vuejs/pinia/commit/44f0762f93bf6ea389dd47574cbf4172989f616b) 11 days ago, but it's not published.

Author: Well maybe not. I see that [vi is just a reference to vitest](https://github.com/vitest-dev/vitest/blob/c1fa19dfc2ccf6c27b2289dc829ec29c53ec66b6/packages/vitest/src/integrations/utils.ts#L86), so either of those symbols should work. I'll have to find another reason why this isn't working on my machine.

(Incidentally, I can see the `vi` global in my test file, but still get the error when calling `createTestingPinia`.

Author: ok @posva I'm sorry for all the noise here, but yes, the reason that you still need to supply `createSpy` with the released version of @pinia/testing (0.0.9) is that support for detecting `vi`/`vitest` was [only added 11 days ago,](e4638ef34c16487b2f35b0de09800f6dcfd5a980) and there has not been a release of @pinia/testing since then.

I'm unfamiliar with your release schedule so perhaps that is intentional. 

I'll prepare this PR to update the documentation in anticipation of a new release being available (it looks like there are comments in the code along the lines of what you'd like the docs to say so I'll use those as a guide).

Reviewer: Sure! I probably haven’t released because there are a few things to change before
